store/postgres/src/relational_queries.rs

```diff
@@ -1291,24 +1293,50 @@ impl<'a> QueryFragment<Pg> for InsertQuery<'a> {
         out.push_sql(self.table.qualified_name.as_str());
 
         out.push_sql("(");
-        for column in self.table.columns.iter() {
-            if self.entity.contains_key(&column.field) {
-                out.push_identifier(column.name.as_str())?;
-                out.push_sql(", ");
+
+        // Build the column name list using the subset of all keys among our entities.
+        let unique_column_names: Vec<&Column> = {
+            let mut btreemap = BTreeMap::new();
+            for (_key, entity) in self.entities.iter() {
+                for column in &self.table.columns {
+                    if entity.get(&column.field).is_some() {
+                        btreemap.entry(column.name.as_str()).or_insert(column);
+                    }
+                }
             }
+            btreemap.into_iter().map(|(_key, value)| value).collect()
+        };
+        for &column in unique_column_names.iter() {
+            out.push_identifier(column.name.as_str())?;
+            out.push_sql(", ");
         }
         out.push_identifier(BLOCK_RANGE_COLUMN)?;
 
-        out.push_sql(")\nvalues(");
-        for column in self.table.columns.iter() {
-            if let Some(value) = self.entity.get(&column.field) {
-                QueryValue(value, &column.column_type).walk_ast(out.reborrow())?;
+        out.push_sql(") values\n");
+
+        // Use a `Peekable` iterator to help us decide how to finalize each line.
+        let mut iter = self.entities.iter().map(|(_key, entity)| entity).peekable();
```

Comment:
Nice! I used to do this with an explicit `first` boolean flag, but a peekable iterator seems nicer.